### PR TITLE
Fix merge write disposition for pyarrow and ClickHouse

### LINKED ISSUE #1972: Replace without cursor then merge with cursor fails on Clickhouse
### dlt version

1.2.0

### Describe the problem

On Clickhouse, If I switch to `write_disposition="merge"` and add a cursor after the pipeline was ran with `write_disposition="replace"` and no cursor, the run will fail with the following error:
```
dlt.destinations.sql_jobs.SqlJobCreationException: Could not create SQLFollowupJob with exception Merge sql job for dataset name `my_database`.`repro_dataset`, staging dataset name `my_database`.`repro_dataset_staging` COULD NOT BE GENERATED. Merge will not be performed. Data for the following tables (['jobs']) is loaded to staging dataset. You may need to write your own materialization. The reason is:
No `row_key` or `unique` column (e.g. `_dlt_id`) in table `jobs`.. Table chain: - columns:
    number:
      name: number
      nullable: false
      data_type: bigint
      primary_key: true
    name:
      name: name
      nullable: true
      data_type: text
  write_disposition: merge
  name: jobs
  resource: jobs
  x-normalizer:
    seen-data: true
```

After this failure mode, subsequent runs will log `The pipeline `run` method will now load the pending load packages. The data you passed to the run function will not be loaded. In order to do that you must run the pipeline again` and attempt to re-run the failed pipeline, failing indefinitely until I run `rm -rf ~/.dlt/pipelines/`.

This does not happen with postgres, and it does not happen if the initial `write_disposition="replace"` was ran with a cursor.

### Expected behavior

I expect to be able to add a cursor and change write disposition on clickhouse in the same way it is possible with postgres.

### Steps to reproduce

Clone and install dependencies for https://github.com/filipesilva/dlt-replace-merge-bug

```
# terminal 1
git clone https://github.com/filipesilva/dlt-replace-merge-bug
cd dlt-replace-merge-bug
python3 -m venv ./env
source .env/bin/activate
```

In separate terminal windows, run the following scripts to launch clickhouse and postgres docker containers:
```
# terminal 2
./scripts/start-postgres.sh
# terminal 3
./scripts/start-clickhouse.sh
```

Go back to the first terminal to and run the postgres pipeline
```
# terminal 1
python sql_database_pipeline.py postgres
```

You should see
```
[I] (.env) filipesilva@Filipes-MacBook-Pro ~/s/dlt-replace-merge-bug (master)> python sql_database_pipeline.py postgres

=== Replace with cursor then merge with cursor ===

Pipeline repro_pipeline load step completed in 1.08 seconds
1 load package(s) were loaded to destination postgres and into dataset repro_dataset
The postgres destination used postgresql://postgres:***@localhost:5432/postgres location to store data
Load package 1729522139.2506921 is LOADED and contains no failed jobs
Pipeline repro_pipeline load step completed in ---
0 load package(s) were loaded to destination postgres and into dataset None
The postgres destination used postgresql://postgres:***@localhost:5432/postgres location to store data

=== Replace without cursor then merge with cursor ===

Pipeline repro_pipeline load step completed in 1.08 seconds
1 load package(s) were loaded to destination postgres and into dataset repro_dataset
The postgres destination used postgresql://postgres:***@localhost:5432/postgres location to store data
Load package 1729522140.885526 is LOADED and contains no failed jobs
Pipeline repro_pipeline load step completed in 2.09 seconds
1 load package(s) were loaded to destination postgres and into dataset repro_dataset
The postgres destination used postgresql://postgres:***@localhost:5432/postgres location to store data
Load package 1729522142.442627 is LOADED and contains no failed jobs
```

But if you run the clickhouse pipeline, it will error out
```
# terminal 1
python sql_database_pipeline.py clickhouse
```
```
[I] (.env) filipesilva@Filipes-MacBook-Pro ~/s/dlt-replace-merge-bug (master)> python sql_database_pipeline.py clickhouse
2024-10-21 15:49:11,040|[WARNING]|48796|8290169408|dlt|pipeline.py|_state_to_props:1603|The destination dlt.destinations.postgres:postgres in state differs from destination dlt.destinations.clickhouse:clickhouse in pipeline and will be ignored

=== Replace with cursor then merge with cursor ===

Pipeline repro_pipeline load step completed in 1.06 seconds
1 load package(s) were loaded to destination clickhouse and into dataset repro_dataset
The clickhouse destination used clickhouse://username:***@localhost:9000/my_database location to store data
Load package 1729522151.1919942 is LOADED and contains no failed jobs
Pipeline repro_pipeline load step completed in ---
0 load package(s) were loaded to destination clickhouse and into dataset None
The clickhouse destination used clickhouse://username:***@localhost:9000/my_database location to store data

=== Replace without cursor then merge with cursor ===

Pipeline repro_pipeline load step completed in 1.06 seconds
1 load package(s) were loaded to destination clickhouse and into dataset repro_dataset
The clickhouse destination used clickhouse://username:***@localhost:9000/my_database location to store data
Load package 1729522152.812721 is LOADED and contains no failed jobs
Traceback (most recent call last):
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/destinations/sql_jobs.py", line 76, in from_table_chain
    for stmt in cls.generate_sql(table_chain, sql_client, params)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/destinations/sql_jobs.py", line 172, in generate_sql
    return cls.gen_merge_sql(table_chain, sql_client)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/destinations/sql_jobs.py", line 538, in gen_merge_sql
    cls._get_row_key_col(table_chain, sql_client, root_table)
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/destinations/sql_jobs.py", line 429, in _get_row_key_col
    col = cls._get_prop_col_or_raise(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/destinations/sql_jobs.py", line 473, in _get_prop_col_or_raise
    raise exception
dlt.destinations.exceptions.MergeDispositionException: Merge sql job for dataset name `my_database`.`repro_dataset`, staging dataset name `my_database`.`repro_dataset_staging` COULD NOT BE GENERATED. Merge will not be performed. Data for the following tables (['jobs']) is loaded to staging dataset. You may need to write your own materialization. The reason is:
No `row_key` or `unique` column (e.g. `_dlt_id`) in table `jobs`.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/load/load.py", line 349, in create_followup_jobs
    if follow_up_jobs := client.create_table_chain_completed_followup_jobs(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/destinations/job_client_impl.py", line 255, in create_table_chain_completed_followup_jobs
    jobs.extend(self._create_merge_followup_jobs(table_chain))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/destinations/impl/clickhouse/clickhouse.py", line 227, in _create_merge_followup_jobs
    return [ClickHouseMergeJob.from_table_chain(table_chain, self.sql_client)]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/destinations/sql_jobs.py", line 82, in from_table_chain
    raise SqlJobCreationException(e, table_chain) from e
dlt.destinations.sql_jobs.SqlJobCreationException: Could not create SQLFollowupJob with exception Merge sql job for dataset name `my_database`.`repro_dataset`, staging dataset name `my_database`.`repro_dataset_staging` COULD NOT BE GENERATED. Merge will not be performed. Data for the following tables (['jobs']) is loaded to staging dataset. You may need to write your own materialization. The reason is:
No `row_key` or `unique` column (e.g. `_dlt_id`) in table `jobs`.. Table chain: - columns:
    number:
      name: number
      nullable: false
      data_type: bigint
      primary_key: true
    name:
      name: name
      nullable: true
      data_type: text
  write_disposition: merge
  name: jobs
  resource: jobs
  x-normalizer:
    seen-data: true


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/pipeline/pipeline.py", line 576, in load
    runner.run_pool(load_step.config, load_step)
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/common/runners/pool_runner.py", line 89, in run_pool
    while _run_func():
          ^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/common/runners/pool_runner.py", line 82, in _run_func
    run_metrics = run_f.run(cast(TExecutor, pool))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/load/load.py", line 638, in run
    self.load_single_package(load_id, schema)
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/load/load.py", line 571, in load_single_package
    running_jobs, finalized_jobs, new_pending_exception = self.complete_jobs(
                                                          ^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/load/load.py", line 446, in complete_jobs
    self.create_followup_jobs(load_id, state, job, schema)
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/load/load.py", line 354, in create_followup_jobs
    raise TableChainFollowupJobCreationFailedException(
dlt.load.exceptions.TableChainFollowupJobCreationFailedException: Failed creating table chain followup jobs for table chain with root table jobs.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/sql_database_pipeline.py", line 50, in <module>
    info = pipeline.run(source, write_disposition="merge")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/pipeline/pipeline.py", line 223, in _wrap
    step_info = f(self, *args, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/pipeline/pipeline.py", line 272, in _wrap
    return f(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/pipeline/pipeline.py", line 715, in run
    return self.load(destination, dataset_name, credentials=credentials)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/pipeline/pipeline.py", line 223, in _wrap
    step_info = f(self, *args, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/pipeline/pipeline.py", line 163, in _wrap
    return f(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/pipeline/pipeline.py", line 272, in _wrap
    return f(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/filipesilva/sandbox/dlt-replace-merge-bug/.env/lib/python3.12/site-packages/dlt/pipeline/pipeline.py", line 583, in load
    raise PipelineStepFailed(
dlt.pipeline.exceptions.PipelineStepFailed: Pipeline execution failed at stage load when processing package 1729522154.360807 with exception:

<class 'dlt.load.exceptions.TableChainFollowupJobCreationFailedException'>
Failed creating table chain followup jobs for table chain with root table jobs.
```

After this failure mode, subsequent runs will log `The pipeline `run` method will now load the pending load packages. The data you passed to the run function will not be loaded. In order to do that you must run the pipeline again` and attempt to re-run the failed pipeline, failing indefinitely until I run `rm -rf ~/.dlt/pipelines/`. You can run this via `./scripts/clear-pending-pipelines.sh`.

### Operating system

macOS

### Runtime environment

Local

### Python version

3.11

### dlt data source

SQLite

### dlt destination

_No response_

### Other deployment details

I'm using Clickhouse as a destination, and Python 3.12.5 but they do not appear on the bug issue dropdowns.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by burnash**
Attempt to fix #1972.

- Extends SqlMergeFollowupJob._get_row_key_col() to fallback to a single primary key if no `row_key` or `unique` columns are found.
- Adds tests for pyarrow load with merge to ClickHouse

### CODE REVIEW

File: tests/load/pipeline/test_clickhouse.py [hunk 1]
```diff
@@ -1,20 +1,37 @@
...
+    assert_records_as_set,
+)
+
+
+@pytest.fixture
+def pipeline(destination_config: DestinationTestConfiguration) -> Generator[Any, None, None]:
+    pipeline = destination_config.setup_pipeline(f"clickhouse_{uniq_id()}", dev_mode=True)
+    yield pipeline
+    with pipeline.sql_client() as client:
+        client.drop_dataset()
```
Comment (rudolfix) [inline tests/load/pipeline/test_clickhouse.py:24]: why you need this? we already have a fixture that drops the dataset (also the staging one) of a pipeline. we run it after every test

File: tests/load/pipeline/test_clickhouse.py [hunk 2]
```diff
@@ -78,3 +97,135 @@ def items2() -> Iterator[TDataItem]:
...
         with pipeline.sql_client() as client:
             client.drop_dataset()
+
+
+@pytest.mark.parametrize(
+    "destination_config",
+    destinations_configs(default_sql_configs=True, subset=["clickhouse"]),
+    ids=lambda x: x.name,
+)
+def test_clickhouse_destination_merge_dicts(
```
Comment (rudolfix) [inline tests/load/pipeline/test_clickhouse.py:107]: please move that test to `test_merge_disposition` and configure it like any other test there. also run with both merge strategies. we need to makes sure we do not have similar problems elsewhere

**Review by rudolfix — CHANGES_REQUESTED** [review]
code LGTM! but we need to move test and run it for all destinations

File: tests/load/pipeline/test_clickhouse.py (hunk 1 above)
Comment (burnash) [reply tests/load/pipeline/test_clickhouse.py:24]: Yes, I just carried this over from the other test thinking that there's something special about ClickHouse. Will adjust.

Comment (burnash) [reply tests/load/pipeline/test_clickhouse.py]: Fixed in b74deb790239b54fa53c9a02af9d9c981a591cd0

**Review by rudolfix — APPROVED** [review]
LGTM

**Pull request merged**